bindings/sundials4py/idas/idas_usersupplied.hpp

```diff
@@ -0,0 +1,497 @@
+ * Programmer(s): Cody J. Balos @ LLNL
+ *------------------------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2025-2025, Lawrence Livermore National Security,
+ * University of Maryland Baltimore County, and the SUNDIALS contributors.
+ * Copyright (c) 2013-2025, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * Copyright (c) 2002-2013, Lawrence Livermore National Security.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ *----------------------------------------------------------------------------*/
+
+#ifndef _SUNDIALS4PY_IDAS_USERSUPPLIED_HPP
+#define _SUNDIALS4PY_IDAS_USERSUPPLIED_HPP
+
+#include <idas/idas.h>
+#include <idas/idas_ls.h>
+
+#include <sundials/sundials_core.hpp>
+
+#include "sundials4py_helpers.hpp"
+
+///////////////////////////////////////////////////////////////////////////////
+// IDAS user-supplied function table
+// Every integrator-level user-supplied function must be in this table.
+// The user-supplied function table is passed to IDAS as user_data.
+///////////////////////////////////////////////////////////////////////////////
+
+struct idas_user_supplied_fn_table
+{
+  // user-supplied function pointers
+  nb::object res, rootfn, ewtn, rwtn, resNLS;
+
+  // idas_ls user-supplied function pointers
+  nb::object lsjacfn, lsprecsetupfn, lsprecsolvefn, lsjactimessetupfn,
+    lsjactimesvecfn, lsjacresfn;
+
+  // idas quadrature user-supplied function pointers
+  nanobind::object resQ, resQS;
+
+  // idas FSA user-supplied function pointers
+  nanobind::object resS;
+};
+
+inline idas_user_supplied_fn_table* idas_user_supplied_fn_table_alloc()
+{
+  // We must use malloc since IDAFree calls free
+  auto fn_table = static_cast<idas_user_supplied_fn_table*>(
+    std::malloc(sizeof(idas_user_supplied_fn_table)));
+
+  // Zero out the memory
+  std::memset(fn_table, 0, sizeof(idas_user_supplied_fn_table));
+
+  return fn_table;
+}
+
+inline idas_user_supplied_fn_table* get_idas_fn_table(void* ida_mem)
+{
+  auto mem      = static_cast<IDAMem>(ida_mem);
+  auto fn_table = static_cast<idas_user_supplied_fn_table*>(mem->python);
+  if (!fn_table)
+  {
+    throw sundials4py::null_function_table(
+      "Failed to get Python function table from IDAS memory");
+  }
+  return fn_table;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// IDAS user-supplied functions
+///////////////////////////////////////////////////////////////////////////////
+
+template<typename... Args>
+inline int idas_res_wrapper(Args... args)
+{
+  return sundials4py::user_supplied_fn_caller<
+    std::remove_pointer_t<IDAResFn>, idas_user_supplied_fn_table, IDAMem,
+    1>(&idas_user_supplied_fn_table::res, std::forward<Args>(args)...);
+}
```

Comment:
```suggestion
inline int idas_res_wrapper(Args... args)
{
  return sundials4py::user_supplied_fn_caller<
    std::remove_pointer_t<IDAResFn>, idas_user_supplied_fn_table, IDAMem,
    1>(&idas_user_supplied_fn_table::res, args...);
}
```
or make args an rvalue reference. Same for remaining functions in file